File: mac/Classes/Model/Actions/Specific/FilterAction.m

#import "FilterAction.h"
#import "Project.h"
#import "LRProjectFile.h"
#import "ScriptInvocationStep.h"
#import "LRFileTargetResult.h"


@implementation FilterAction

- (NSString *)label {
    return self.type.name;
}

- (void)loadFromMemento:(NSDictionary *)memento {
    [super loadFromMemento:memento];

    NSString *inputFilter = self.type.manifest[@"input"];
    self.intrinsicInputPathSpec = [ATPathSpec pathSpecWithString:inputFilter syntaxOptions:ATPathSpecSyntaxFlavorExtended];
}

- (void)updateMemento:(NSMutableDictionary *)memento {
    [super updateMemento:memento];
    //    memento[@"output"] = self.outputFilterOption.memento;
}

- (void)configureStep:(ScriptInvocationStep *)step forFile:(LRProjectFile *)file {
    [super configureStep:step forFile:file];

    [step addFileValue:file forSubstitutionKey:@"src"];
}

- (void)didCompleteCompilationStep:(ScriptInvocationStep *)step forFile:(LRProjectFile *)file {
    LRProjectFile *outputFile = [step fileForKey:@"src"];
    [file.project hackhack_didFilterFile:outputFile];
}

- (void)compileFile:(LRProjectFile *)file inProject:(Project *)project result:(LROperationResult *)result completionHandler:(dispatch_block_t)completionHandler {
    if (![project hackhack_shouldFilterFile:file]) {
        completionHandler();
        return;
    }

    [super compileFile:file inProject:project result:result completionHandler:completionHandler];
}

- (BOOL)supportsFileTargets {
    return YES;
}

- (LRTargetResult *)fileTargetForRootFile:(LRProjectFile *)sourceFile {
    return [[LRFileTargetResult alloc] initWithAction:self sourceFile:sourceFile];
}


@end

File: mac/Classes/Model/Actions/Specific/FilterAction.swift

import Foundation

@objc(FilterAction)
class FilterAction : ScriptInvocationAction {

    override var label: String {
        return type.name
    }

    override var supportsFileTargets: Bool {
        return true
    }

    override func loadFromMemento(memento: NSDictionary!) {
        super.loadFromMemento(memento)

        let inputFilter = type.manifest["input"]! as String
        intrinsicInputPathSpec = ATPathSpec(string: inputFilter, syntaxOptions: ATPathSpecSyntaxFlavorExtended)
    }

    override func updateMemento(memento: NSMutableDictionary!) {
        super.updateMemento(memento)
    }

    override func fileTargetForRootFile(file: LRProjectFile!) -> LRTarget! {
        return LRFileTarget(action: self, sourceFile: file)
    }

    override func configureStep(step: ScriptInvocationStep, forFile file: LRProjectFile) {
        super.configureStep(step, forFile: file)
        step.addFileValue(file, forSubstitutionKey: "src")
    }

    override func didCompleteCompilationStep(step: ScriptInvocationStep, forFile file: LRProjectFile) {
        let outputFile = step.fileForKey("src")!
        file.project.hackhack_didFilterFile(outputFile)
    }

    override func compileFile(file: LRProjectFile!, inProject project: Project!, result: LROperationResult!, completionHandler: dispatch_block_t!) {
        if (!project.hackhack_shouldFilterFile(file)) {
            completionHandler()
            return
        }

        super.compileFile(file, inProject: project, result: result, completionHandler: completionHandler)
    }

}
